File: agent/internal/behavior/engine.go
package behavior

import (
	"context"
	"fmt"
	"math/rand"
	"time"

	"apc/internal/system"
)

// Engine drives simulated presence; it delegates to a Platform implementation.
type Engine struct {
	platform system.Platform
}

func New(p system.Platform) *Engine {
	return &Engine{platform: p}
}

// Run starts a simple loop that emits placeholder actions and performs micro moves.
func (e *Engine) Run(ctx context.Context, mode string, emit func(action string, meta map[string]string)) {
	ticker := time.NewTicker(500 * time.Millisecond)
	defer ticker.Stop()

	var lastX, lastY int
	havePos := false
	pausedUntil := time.Time{}

	if e.platform != nil {
		if err := e.platform.PreventSleep(true); err != nil {
			emit("prevent_sleep_error", map[string]string{"error": err.Error()})
		}
		defer e.platform.PreventSleep(false)
	}

	emit("engine_start", map[string]string{"mode": mode})

	for {
		select {
		case <-ctx.Done():
			emit("engine_stop", map[string]string{"reason": "context_cancel"})
			return
		case <-ticker.C:
			// Detect user movement and pause if user is active.
			if e.platform != nil {
				if x, y, err := e.platform.CursorPos(); err == nil {
					if havePos {
						if dist(x-lastX, y-lastY) > 80 { // user activity threshold
							pausedUntil = time.Now().Add(15 * time.Second)
							emit("user_active_pause", map[string]string{"until": pausedUntil.UTC().Format(time.RFC3339)})
						}
					}
					lastX, lastY, havePos = x, y, true
				}
			}

			if !pausedUntil.IsZero() && time.Now().Before(pausedUntil) {
				continue
			}

			dx := rand.Intn(201) - 100 // range -100..100
			dy := rand.Intn(201) - 100
			if dx == 0 && dy == 0 {
				dx = 20 // ensure visible movement
			}
			if e.platform != nil {
				if err := e.platform.SimulateMove(dx, dy); err != nil {
					emit("move_error", map[string]string{"error": err.Error()})
				} else {
					// update last position after our move
					if x, y, err := e.platform.CursorPos(); err == nil {
						lastX, lastY, havePos = x, y, true
					}
				}
			}
			emit("micro_move", map[string]string{"mode": mode, "dx": intToStr(dx), "dy": intToStr(dy)})
		}
	}
}

func dist(dx, dy int) int {
	if dx < 0 {
		dx = -dx
	}
	if dy < 0 {
		dy = -dy
	}
	return dx + dy
}

func intToStr(v int) string {
	return fmt.Sprintf("%d", v)
}

File: agent/internal/behavior/engine.py
"""
Behavior engine that simulates user presence with cursor movements.
Detects user activity and adapts accordingly.
"""

import logging
import random
import threading
import time

logger = logging.getLogger(__name__)


class Engine:
    """Drives simulated presence with human-like micro movements."""

    def __init__(self, platform):
        self.platform = platform

    def run(self, cancel_event, mode, emit):
        """
        Run the behavior engine until cancel_event is set.

        Args:
            cancel_event: threading.Event to signal stop
            mode: Behavior mode (e.g., "Meeting", "Build")
            emit: Callback function(action, meta) for logging
        """
        interval = 0.5  # 500ms between movements

        last_x = 0
        last_y = 0
        have_pos = False
        paused_until = None

        if self.platform:
            try:
                self.platform.prevent_sleep(True)
            except Exception as e:
                emit("prevent_sleep_error", {"error": str(e)})

        emit("engine_start", {"mode": mode})

        ticker = threading.Event()
        last_tick = time.time()

        try:
            while not cancel_event.is_set():
                now = time.time()
                elapsed = now - last_tick

                if elapsed >= interval:
                    last_tick = now

                    # Detect user movement and pause if user is active
                    if self.platform:
                        try:
                            x, y = self.platform.cursor_pos()
                            if have_pos:
                                dist = abs(x - last_x) + abs(y - last_y)
                                if dist > 80:  # User activity threshold
                                    paused_until = time.time() + 15  # Pause for 15 seconds
                                    emit(
                                        "user_active_pause",
                                        {"until": paused_until},
                                    )
                            last_x, last_y, have_pos = x, y, True
                        except Exception as e:
                            logger.debug(f"Could not get cursor position: {e}")

                    # Check if paused
                    if paused_until and time.time() < paused_until:
                        time.sleep(0.05)
                        continue

                    # Generate random movement
                    dx = random.randint(-100, 100)
                    dy = random.randint(-100, 100)

                    if dx == 0 and dy == 0:
                        dx = 20  # Ensure visible movement

                    if self.platform:
                        try:
                            self.platform.simulate_move(dx, dy)
                            # Update position after move
                            try:
                                x, y = self.platform.cursor_pos()
                                last_x, last_y, have_pos = x, y, True
                            except Exception:
                                pass
                        except Exception as e:
                            emit("move_error", {"error": str(e)})
                    else:
                        # Cloud mode - no actual movement
                        pass

                    emit("micro_move", {"mode": mode, "dx": str(dx), "dy": str(dy)})
                else:
                    time.sleep(0.01)  # Small sleep to avoid busy-waiting

        except Exception as e:
            logger.error(f"Engine error: {e}")
            emit("engine_error", {"error": str(e)})
        finally:
            if self.platform:
                try:
                    self.platform.prevent_sleep(False)
                except Exception as e:
                    logger.debug(f"Could not disable sleep prevention: {e}")

            emit("engine_stop", {"reason": "context_cancel"})
